### PR TITLE
Feature/43 implement create drop schema

### PR DESCRIPTION
**Pull request by AnastasiiaSergienko**

### CODE REVIEW

File: src/main/java/com/exasol/sql/ddl/drop/DropSchema.java [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+        return this;
+    }
+
+    /**
+     * Add CASCADE clause into a DROP SCHEMA statement
+     *
+     * @return <code>this</code> for fluent programming
+     */
+    public synchronized DropSchema cascade() {
+        cascade = new Cascade(this);
```
Comment (redcatbear) [inline src/main/java/com/exasol/sql/ddl/drop/DropSchema.java]: Letting both `cascade()` and `restrict()` return a `DropSchema` allows to use the mutually exclusive options together at compile time.

There are possible alternatives:

1. Introduce a "terminator" class and return an object of the class that prevents further chaining. In this case the `restrict()` and `cascade()` must be the last methods in a chain.
1. Introduce two separate classes that are given back. While this is convenient for the end users, it creates a lot of duplication and does not scale if the chain gets more complex.
1. Return `void`

File: src/main/java/com/exasol/sql/ddl/drop/DropSchema.java [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+     *
+     * @return <code>this</code> for fluent programming
+     */
+    public synchronized DropSchema restrict() {
+        restrict = new Restrict(this);
+        return this;
+    }
+
+    /**
+     * Get a schema name
```
Comment (redcatbear) [inline src/main/java/com/exasol/sql/ddl/drop/DropSchema.java]: "the"

File: src/main/java/com/exasol/sql/ddl/drop/DropSchema.java [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+     *
+     * @return <code>this</code> for fluent programming
+     */
+    public synchronized DropSchema cascade() {
+        cascade = new Cascade(this);
+        return this;
+    }
+
+    /**
+     * Add RESTRICT clause into a DROP SCHEMA statement
```
Comment (redcatbear) [inline src/main/java/com/exasol/sql/ddl/drop/DropSchema.java]: "into" --> "to"

File: src/main/java/com/exasol/sql/ddl/drop/DropSchema.java [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+    /**
+     * Get a schema name
+     *
+     * @return schema name
+     */
+    public String getSchemaName() {
+        return schema.getName();
+    }
+
+    public Cascade getCascade() {
```
Comment (redcatbear) [inline src/main/java/com/exasol/sql/ddl/drop/DropSchema.java]: Add missing JavaDoc.

File: src/main/java/com/exasol/sql/ddl/drop/DropSchema.java [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+    /**
+     * Get true when IF EXISTS clause presents
+     *
+     * @return if exists
+     */
+    public boolean getIfExists() {
+        return ifExists;
+    }
+
+    private void validateCascadeAndRestrict() {
```
Comment (redcatbear) [inline src/main/java/com/exasol/sql/ddl/drop/DropSchema.java]: This is not the right place for a validation. We should do all validations in a separate visitor to have them all in one place.

File: src/test/java/com/exasol/hamcrest/SqlFragmentRenderResultMatcher.java [hunk 6]
```diff
@@ -54,6 +57,14 @@ public boolean matchesSafely(final Fragment fragment) {
             final DropTableRenderer renderer = new DropTableRenderer(this.config);
             ((DropTableFragment) root).accept(renderer);
             this.renderedText = renderer.render();
+        } else if (root instanceof CreateSchemaFragment) {
```
Comment (redcatbear) [inline src/test/java/com/exasol/hamcrest/SqlFragmentRenderResultMatcher.java]: We are approaching the point where this part should be done via reflection. This method contains a whole lot of tight coupling.

File: src/test/java/com/exasol/sql/ddl/drop/TestDropSchema.java [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+        dropSchema = StatementFactory.getInstance().dropSchema(TEST_SCHEMA_NAME);
+    }
+
+    @Test
+    void getTableName() {
+        assertThat(this.dropSchema.getSchemaName(), equalTo(TEST_SCHEMA_NAME));
+    }
+
+    @Test
+    void ifExists() {
```
Comment (redcatbear) [inline src/test/java/com/exasol/sql/ddl/drop/TestDropSchema.java]: Split into two tests please. One test the default setting, the other one the setter. This way you have only one assert in each.

File: src/test/java/com/exasol/sql/ddl/drop/TestDropSchema.java [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+        assertFalse(this.dropSchema.getIfExists());
+        this.dropSchema.ifExists();
+        assertTrue(this.dropSchema.getIfExists());
+    }
+
+    @Test
+    void setCascade() {
+        assertNull(this.dropSchema.getCascade());
```
Comment (redcatbear) [inline src/test/java/com/exasol/sql/ddl/drop/TestDropSchema.java]: Same here.

File: src/test/java/com/exasol/sql/ddl/drop/TestDropSchema.java [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+        this.dropSchema.cascade();
+        assertNotNull(this.dropSchema.getCascade());
+    }
+
+    @Test
+    void setRestrict() {
+        assertNull(this.dropSchema.getRestrict());
```
Comment (redcatbear) [inline src/test/java/com/exasol/sql/ddl/drop/TestDropSchema.java]: And here.

File: src/main/java/com/exasol/sql/ddl/Schema.java [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+     * @param root       SQL statement this schema belongs to
+     * @param schemaName schema name
+     */
+    public Schema(Fragment root, String schemaName) {
+        super(root);
+        this.name = schemaName;
+    }
+
+    /**
+     * Get a schema name
```
[the remark this answers is not in the document]
Comment (AnastasiiaSergienko) [reply src/main/java/com/exasol/sql/ddl/Schema.java]: fixed

File: src/main/java/com/exasol/sql/ddl/create/CreateTable.java [hunk 11]
```diff
@@ -151,11 +149,21 @@ public synchronized CreateTable intervalYearToMonthColumn(final String columnNam
         return this;
     }
 
-    protected String getTableName() {
+    /**
+     * Get a table name
```
[the remark this answers is not in the document]
Comment (AnastasiiaSergienko) [reply src/main/java/com/exasol/sql/ddl/create/CreateTable.java]: fixed

File: src/main/java/com/exasol/sql/ddl/create/CreateTable.java [hunk 12]
```diff
@@ -151,11 +149,21 @@ public synchronized CreateTable intervalYearToMonthColumn(final String columnNam
...
+     *
+     * @return table name
+     */
+    public String getTableName() {
         return this.table.getName();
     }
 
-    protected ColumnsDefinition getColumns() {
+    /**
+     * Get columns of a table
```
[the remark this answers is not in the document]
Comment (AnastasiiaSergienko) [reply src/main/java/com/exasol/sql/ddl/create/CreateTable.java]: fixed

File: src/main/java/com/exasol/sql/ddl/drop/CascadeConstraints.java [hunk 13]
```diff
@@ -4,7 +4,7 @@
 import com.exasol.sql.Fragment;
 
 /**
- * This class represents "cascade constraints" expression in DROP TABLE SQL statement
+ * This class represents CASCADE CONSTRAINTS clause in DROP TABLE SQL statement
```
[the remark this answers is not in the document]
Comment (AnastasiiaSergienko) [reply src/main/java/com/exasol/sql/ddl/drop/CascadeConstraints.java]: fixed

File: src/main/java/com/exasol/sql/ddl/drop/DropSchema.java (hunk 1 above)
Comment (AnastasiiaSergienko) [reply src/main/java/com/exasol/sql/ddl/drop/DropSchema.java]: fixed

File: src/main/java/com/exasol/sql/ddl/drop/DropSchema.java (hunk 2 above)
Comment (AnastasiiaSergienko) [reply src/main/java/com/exasol/sql/ddl/drop/DropSchema.java]: fixed

File: src/main/java/com/exasol/sql/ddl/drop/DropSchema.java (hunk 3 above)
Comment (AnastasiiaSergienko) [reply src/main/java/com/exasol/sql/ddl/drop/DropSchema.java]: fixed

File: src/main/java/com/exasol/sql/ddl/drop/DropSchema.java (hunk 4 above)
Comment (AnastasiiaSergienko) [reply src/main/java/com/exasol/sql/ddl/drop/DropSchema.java]: added

File: src/main/java/com/exasol/sql/ddl/drop/DropSchema.java (hunk 5 above)
Comment (AnastasiiaSergienko) [reply src/main/java/com/exasol/sql/ddl/drop/DropSchema.java]: Ticket created #45

File: src/test/java/com/exasol/sql/ddl/drop/TestDropSchema.java (hunk 7 above)
Comment (AnastasiiaSergienko) [reply src/test/java/com/exasol/sql/ddl/drop/TestDropSchema.java]: fixed

File: src/test/java/com/exasol/sql/ddl/drop/TestDropSchema.java (hunk 8 above)
Comment (AnastasiiaSergienko) [reply src/test/java/com/exasol/sql/ddl/drop/TestDropSchema.java]: fixed

File: src/test/java/com/exasol/sql/ddl/drop/TestDropSchema.java (hunk 9 above)
Comment (AnastasiiaSergienko) [reply src/test/java/com/exasol/sql/ddl/drop/TestDropSchema.java]: fixed

File: src/test/java/com/exasol/hamcrest/SqlFragmentRenderResultMatcher.java (hunk 6 above)
Comment (AnastasiiaSergienko) [reply src/test/java/com/exasol/hamcrest/SqlFragmentRenderResultMatcher.java]: done

File: src/test/java/com/exasol/hamcrest/SqlFragmentRenderResultMatcher.java [hunk 14]
```diff
@@ -79,12 +38,57 @@ public static SqlFragmentRenderResultMatcher rendersTo(final String expectedText
...
+        } catch (final ClassNotFoundException | InstantiationException | IllegalAccessException | NoSuchMethodException
+                | InvocationTargetException cause) {
+            throw new UnsupportedOperationException(
+                    "Don't know how to render fragment of type\"" + rootClassName + "\".", cause);
+        }
+        return this.renderedText.equals(this.expectedText);
+    }
+
+    private String getRendererClassPath(final String oldClassPath) {
+        final String[] split = oldClassPath.split("\\.");
```
Comment (redcatbear) [inline src/test/java/com/exasol/hamcrest/SqlFragmentRenderResultMatcher.java]: `split` -> `pathElements`

**Pull request merged**